README.md
# <img src="https://github.com/pip-services/pip-services/raw/master/design/Logo.png" alt="Pip.Services Logo" style="max-width:30%"> <br/> Mappings microservice

This is Mappings microservice from Pip.Services library. 
It stores Mappings between internally and external service

The microservice currently supports the following deployment options:
* Deployment platforms: Standalone Process, Seneca
* External APIs: HTTP/REST
* Persistence: Flat Files, MongoDB

This microservice has no dependencies on other microservices.

<a name="links"></a> Quick Links:

* [Download Links](doc/Downloads.md)
* [Development Guide](doc/Development.md)
* [Configuration Guide](doc/Configuration.md)
* [Deployment Guide](doc/Deployment.md)
* Mappings Service
  - [Node.js SDK](https://github.com/pip-services-integration2/service-mappings-node)
* Communication Protocols
  - [HTTP Version 1](doc/HttpProtocolV1.md)
 
## Contract

Logical contract of the microservice is presented below. For physical implementation (HTTP/REST, Thrift, Seneca, Lambda, etc.),
please, refer to documentation of the specific protocol.

```typescript
class MappingV1 implements IStringIdentifiable {
    
    public id:string;
    public collection: string;
    public internal_id: string;
    public external_id: string;
    public expiration_time: Date;

}

interface IMappingsController {
    getCollectionNames(correlationId: string): Promise<Array<string>>;
    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams): Promise<DataPage<MappingV1>>;
    addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number): Promise<void>;
    mapToExternal(correlationId: string, collection: string, internalId: string): Promise<string>;
    mapToInternal(correlationId: string, collection: string, externalId: string): Promise<string>;
    deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string): Promise<void>;
    deleteExpiredMappings(correlationId: string): Promise<void>;
}
```

## Download

Right now the only way to get the microservice is to check it out directly from github repository
```bash
git clone git@github.com:pip-services-integration2/client-mappings-node.git
```

Pip.Service team is working to implement packaging and make stable releases available for your 
as zip downloadable archieves.

## Run

Add **config.yml** file to the root of the microservice folder and set configuration parameters.
As the starting point you can use example configuration from **config.example.yml** file. 

Example of microservice configuration
```yaml
- descriptor: "pip-services-container:container-info:default:default:1.0"
  name: "service-mappings"
  description: "Mappings microservice"

- descriptor: "pip-services-commons:logger:console:default:1.0"
  level: "trace"

- descriptor: "service-mappings:persistence:file:default:1.0"
  path: "./data/mappings.json"

- descriptor: "service-mappings:controller:default:default:1.0"

- descriptor: "service-mappings:service:http:default:1.0"
  connection:
    protocol: "http"
    host: "0.0.0.0"
    port: 8080
```
 
For more information on the microservice configuration see [Configuration Guide](Configuration.md).

Start the microservice using the command:
```bash
node run
```

## Install

The easiest way to work with the microservice is to use client SDK. 
The complete list of available client SDKs for different languages is listed in the [Quick Links](#links)

If you use Node.js then you should add dependency to the client SDK into **package.json** file of your project
```javascript
{
    ...
    "dependencies": {
        ....
        "client-mappings-node": "^1.0.*",
        ...
    }
}
```


## Use

Inside your code get the reference to the client SDK
```typescript
 import { MappingsHttpClientV1 } from 'client-mappings-node';
```

Define client configuration parameters.

```typescript
// Client configuration
let httpConfig = ConfigParams.fromTuples(
    "connection.protocol", "http",
    "connection.host", "localhost",
    "connection.port", 3000
);
client.configure(httpConfig);
```

Instantiate the client and open connection to the microservice
```typescript
// Create the client instance
client = new MappingsHttpClientV1();

// Connect to the microservice
await client.open(null);

// Work with the microservice
...
```
Now the client is ready to perform operations:

Add mapping:
```typescript 
let mapping = await client.addMapping("123", "Common.Collection", "123", "789", 60 * 1000);
console.dir('Mapping was created successfull');
```

Get mappings by filter:
```typescript    

let page = await client.getMappings("123", new FilterParams(), new PagingParams());

console.dir('Mappings was recived successfull');
for (let mapping in page.data) {
    console.dir('Mappings:');
    console.dir(mapping.toString());
}

page = await client.getMappings(null, FilterParams.fromTuples("collection", "Common.Collection"), new PagingParams(1, 10, false));

console.dir('Mappings was recived successfull');
for (let mapping in page.data) {
    console.dir('Mappings:');
    console.dir(mapping.toString());
}
```

Delete mapping:
```typescript    

client.deleteMappingById("123", "Common.Collection", "123", "789");

console.dir('Mapping was delete successfull');
```

## Control functions

Get collection names:
```typescript
let collections = await client.getCollectionNames("123");
console.dir('Collections names:');
console.dir(collections);
```

Get external id by collection name and internal id:
```typescript
let id = await client.mapToExternal("123", "Common.Collection", "123");

console.dir('External id:', id);
```

Get internal id by collection name and external id:
```typescript
let id = await client.mapToInternal("123", "Common.Collection", "789"); 

console.dir('Internal id:', id);
```


## Acknowledgements

This microservice was created and currently maintained by 
- *Sergey Seroukhov*.
- *Levichev Dmitry*


CHANGELOG.md
# Mappings Client SDK ChangeLog

## <a name="1.0.0"></a> 1.0.0 (2022-08-04)

### Features
* **MongoDB** data access using Mongoose framework
* **File** data access 
* **Memory** data access 
* **REST** API client and service
* **Unit tests** using Mocha and Matcha

### Breaking Changes
No breaking changes since this is the first version

### Bug Fixes
No fixes in this version



test/version1/MappingsClientV1Fixture.ts
const assert = require('chai').assert;

import { FilterParams } from 'pip-services3-commons-nodex';
import { PagingParams } from 'pip-services3-commons-nodex';

import { IMappingsClient } from '../../src/version1/IMappingsClient';

export class MappingsClientV1Fixture {
    private _client: IMappingsClient;

    constructor(persistence) {
        assert.isNotNull(persistence);
        this._client = persistence;
    }

    public async testGetMappingCollections() {
        // Add mappings
        await this._client.addMapping(null, "Common.Collection", "123", "789", 60 * 1000);
        await this._client.addMapping(null, "Common.AnotherCollection", "123", "543", 60 * 1000);
        await this._client.addMapping(null, "Common.Collection", "ABC", "XYZ", 60 * 1000);

        let items = await this._client.getCollectionNames(null);

        assert.equal(2, items.length);
        assert.include(items, "Common.Collection");
        assert.include(items, "Common.AnotherCollection");
    }

    public async testGetMappings() {
        // Add mappings
        await this._client.addMapping(null, "Common.Collection", "123", "789", 60 * 1000);
        await this._client.addMapping(null, "Common.AnotherCollection", "123", "543", 60 * 1000);
        await this._client.addMapping(null, "Common.Collection", "ABC", "XYZ", 60 * 1000);
        await this._client.addMapping(null, "Common.Collection", "AAA", "111", 60 * 1000);

        let mappings = await this._client.getMappings(null, FilterParams.fromTuples("collection", "Common.Collection"), new PagingParams(1, 10, false));

        assert.isNotNull(mappings.data);
        assert.equal(2, mappings.data.length);
    }

    public async testMapping() {
        // Add mappings
        await this._client.addMapping(null, "Common.Collection", "123", "789", 60 * 1000);
        await this._client.addMapping(null, "Common.AnotherCollection", "123", "543", 60 * 1000);
        await this._client.addMapping(null, "Common.Collection", "ABC", "XYZ", 60 * 1000);

        // Test internal mappings
        let id = await this._client.mapToExternal(null, "Common.Collection", "123");
        assert.equal("789", id);

        // Test external mappings
        id = await this._client.mapToInternal(null, "Common.Collection", "789");
        assert.equal("123", id);

        // Test different collection
        id = await this._client.mapToExternal(null, "Common.AnotherCollection", "123");
        assert.equal("543", id);

        // Test non-exiting collection
        id = await this._client.mapToExternal(null, "Common.YetAnotherCollection", "123");
        assert.isNull(id);

        // Test non-exiting mapping
        id = await this._client.mapToExternal(null, "Common.Collection", "555");
        assert.isNull(id);
    }
}


test/version1/MappingsHttpClient.test.ts
import { ConfigParams } from 'pip-services3-commons-nodex';
import { Descriptor } from 'pip-services3-commons-nodex';
import { References } from 'pip-services3-commons-nodex';

import { MappingsMemoryPersistence } from 'service-mappings-node';
import { MappingsController } from 'service-mappings-node';
import { MappingsHttpServiceV1 } from 'service-mappings-node';

import { MappingsHttpClientV1 } from '../../src/version1/MappingsHttpClientV1';
import { MappingsClientV1Fixture } from './MappingsClientV1Fixture';

suite('MappingsHttpClientV1', () => {
    let persistence: MappingsMemoryPersistence;
    let controller: MappingsController;
    let service: MappingsHttpServiceV1;
    let client: MappingsHttpClientV1;
    let fixture: MappingsClientV1Fixture;

    setup(async () => {
        persistence = new MappingsMemoryPersistence();
        persistence.configure(new ConfigParams());

        controller = new MappingsController();
        controller.configure(new ConfigParams());

        let httpConfig = ConfigParams.fromTuples(
            'connection.protocol', 'http',
            'connection.port', 3000,
            'connection.host', 'localhost'
        );

        service = new MappingsHttpServiceV1();
        service.configure(httpConfig);

        client = new MappingsHttpClientV1();
        client.configure(httpConfig);

        let references = References.fromTuples(
            new Descriptor('service-mappings', 'persistence', 'memory', 'default', '1.0'), persistence,
            new Descriptor('service-mappings', 'controller', 'default', 'default', '1.0'), controller,
            new Descriptor('service-mappings', 'service', 'http', 'default', '1.0'), service,
            new Descriptor('service-mappings', 'client', 'http', 'default', '1.0'), client
        );
        controller.setReferences(references);
        service.setReferences(references);
        client.setReferences(references);

        fixture = new MappingsClientV1Fixture(client);

        await persistence.open(null);
        await service.open(null);
        await service.open(null);
        await client.open(null);
    });

    teardown(async () => {
        await client.close(null);
        await service.close(null);
        await persistence.close(null);;
    });

    test('Get Mapping Collections', async () => {
        await fixture.testGetMappingCollections();
    });

    test('Get Mappings', async () => {
        await fixture.testGetMappings();
    });

    test('Mapping', async () => {
        await fixture.testMapping();
    });

});

test/version1/MappingsDirectClientV1.test.ts

import { ConfigParams } from 'pip-services3-commons-nodex';
import { Descriptor } from 'pip-services3-commons-nodex';
import { References } from 'pip-services3-commons-nodex';

import { MappingsMemoryPersistence } from 'service-mappings-node';
import { MappingsController } from 'service-mappings-node';

import { MappingsDirectClientV1 } from '../../src/version1/MappingsDirectClientV1';
import { MappingsClientV1Fixture } from './MappingsClientV1Fixture';


suite('MappingsDirectClientV1', () => {
    let persistence: MappingsMemoryPersistence;
    let controller: MappingsController;
    let client: MappingsDirectClientV1;
    let fixture: MappingsClientV1Fixture;

    setup(async () => {
        persistence = new MappingsMemoryPersistence();
        persistence.configure(new ConfigParams());

        controller = new MappingsController();
        controller.configure(new ConfigParams());

        client = new MappingsDirectClientV1();

        let references = References.fromTuples(
            new Descriptor('service-mappings', 'persistence', 'memory', 'default', '1.0'), persistence,
            new Descriptor('service-mappings', 'controller', 'default', 'default', '1.0'), controller,
            new Descriptor('service-mappings', 'client', 'direct', 'default', '1.0'), client
        );

        controller.setReferences(references);
        client.setReferences(references);

        fixture = new MappingsClientV1Fixture(client);

        await persistence.open(null);
    });

    teardown(async () => {
        await persistence.close(null);
    });

    test('Get Mapping Collections', async () => {
        await fixture.testGetMappingCollections();
    });

    test('Get Mappings', async () => {
        await fixture.testGetMappings();
    });

    test('Mapping', async () => {
        await fixture.testMapping();
    });

});

src/version1/IMappingsClient.ts
import { DataPage } from 'pip-services3-commons-nodex';
import { FilterParams } from 'pip-services3-commons-nodex';
import { PagingParams } from 'pip-services3-commons-nodex';

import { MappingV1 } from './MappingV1';

export interface IMappingsClient {
    getCollectionNames(correlationId: string): Promise<Array<string>>;

    getMappings(correlationId: string, filter: FilterParams, paging: PagingParams): Promise<DataPage<MappingV1>>;

    addMapping(correlationId: string, collection: string,
        internalId: string, externalId: string, timeToLive: number): Promise<void>;

    mapToExternal(correlationId: string, collection: string, internalId: string): Promise<string>;

    mapToInternal(correlationId: string, collection: string, externalId: string): Promise<string>;

    deleteMapping(correlationId: string, collection: string,
        internalId: string, externalId: string): Promise<void>;
}


src/version1/MappingsHttpClientV1.ts
import { IMappingsClient } from "./IMappingsClient";
import { DataPage, FilterParams, PagingParams } from "pip-services3-commons-nodex";
import { CommandableHttpClient } from 'pip-services3-rpc-nodex';
import { MappingV1 } from "service-mappings-node";

export class MappingsHttpClientV1 extends CommandableHttpClient implements IMappingsClient {

    public constructor() {
        super('v1/mappings');
    }

    public async getCollectionNames(correlationId: string): Promise<string[]> {
        let timing = this.instrument(correlationId, 'mappings.get_collection_names');

        try {
            return await this.callCommand(
                'get_collection_names',
                correlationId,
                {}
            );
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async getMappings(correlationId: string, filter: FilterParams, paging: PagingParams): Promise<DataPage<MappingV1>> {
        let timing = this.instrument(correlationId, 'mappings.get_mappings');

        try {
            return await this.callCommand(
                'get_mappings',
                correlationId,
                {
                    filter: filter,
                    paging: paging
                }
            );
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number): Promise<void> {
        let timing = this.instrument(correlationId, 'mappings.add_mapping');

        try {
            return await this.callCommand(
                'add_mapping',
                correlationId,
                {
                    collection: collection,
                    internal_id: internalId,
                    external_id: externalId,
                    ttl: timeToLive
                }
            );
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async mapToExternal(correlationId: string, collection: string, internalId: string): Promise<string> {
        let timing = this.instrument(correlationId, 'mappings.map_to_external');

        try {
            return await this.callCommand(
                'map_to_external',
                correlationId,
                {
                    collection: collection,
                    internal_id: internalId
                }
            );
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async mapToInternal(correlationId: string, collection: string, externalId: string): Promise<string> {
        let timing = this.instrument(correlationId, 'mappings.map_to_internal');

        try {
            return await this.callCommand(
                'map_to_internal',
                correlationId,
                {
                    collection: collection,
                    external_id: externalId
                }
            );
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string): Promise<void> {
        let timing = this.instrument(correlationId, 'mappings.map_to_internal');

        try {
            return await this.callCommand(
                'map_to_internal',
                correlationId,
                {
                    collection: collection,
                    external_id: externalId
                }
            );
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }
}

src/version1/MappingsNullClientV1.ts
import { IMappingsClient } from "./IMappingsClient";
import { FilterParams, PagingParams, DataPage } from "pip-services3-commons-nodex";
import { MappingV1 } from "./MappingV1";


export class MappingsNullClientV1 implements IMappingsClient {
    public async getCollectionNames(correlationId: string): Promise<string[]> {
        return new Array<string>();
    }
    public async getMappings(correlationId: string, filter: FilterParams, paging: PagingParams): Promise<DataPage<MappingV1>> {
        return new DataPage<MappingV1>();
    }
    public async addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number): Promise<void> {
        return null;
    }
    public async mapToExternal(correlationId: string, collection: string, internalId: string): Promise<string> {
        return null;
    }
    public async mapToInternal(correlationId: string, collection: string, externalId: string): Promise<string> {
        return null;
    }
    public async deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string): Promise<void> {
        return null;
    }


}

src/version1/MappingsDirectClientV1.ts
import { IMappingsClient } from "./IMappingsClient";
import { FilterParams, PagingParams, Descriptor, DataPage } from "pip-services3-commons-nodex";
import { DirectClient } from 'pip-services3-rpc-nodex';
import { MappingV1 } from "service-mappings-node";

export class MappingsDirectClientV1 extends DirectClient<any> implements IMappingsClient {
    
    public constructor() {
        super();
        this._dependencyResolver.put('controller', new Descriptor('service-mappings', 'controller', '*', '*', '1.0'));
    }

    public async getCollectionNames(correlationId: string): Promise<string[]> {
        let timing = this.instrument(correlationId, 'mappings.get_collections_names');
        
        try {
            return await this._controller.getCollectionNames(correlationId);
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async getMappings(correlationId: string, filter: FilterParams, paging: PagingParams): Promise<DataPage<MappingV1>> {
        let timing = this.instrument(correlationId, 'mappings.get_mappings');
        
        try {
            return await this._controller.getMappings(correlationId, filter, paging);
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number): Promise<void> {
        let timing = this.instrument(correlationId, 'mappings.add_mapping');
        
        try {
            return await this._controller.addMapping(correlationId, collection, internalId, externalId, timeToLive);
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async mapToExternal(correlationId: string, collection: string, internalId: string): Promise<string> {
        let timing = this.instrument(correlationId, 'mappings.map_to_external');
        
        try {
            return await this._controller.mapToExternal(correlationId, collection, internalId);
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async mapToInternal(correlationId: string, collection: string, externalId: string): Promise<string> {
        let timing = this.instrument(correlationId, 'mappings.map_to_internal');

        try {
            return await this._controller.mapToInternal(correlationId, collection, externalId);
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

    public async deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string): Promise<void> {
        let timing = this.instrument(correlationId, 'mappings.delete_mapping');

        try {
            return await this._controller.deleteMapping(correlationId, collection, internalId, externalId);
        } catch (err) {
            timing.endFailure(err);
            throw err;
        } finally {
            timing.endTiming();
        }
    }

}

src/version1/MappingsMemoryClientV1.ts
import { IMappingsClient } from "./IMappingsClient";
import { FilterParams, PagingParams, DataPage, IdGenerator } from "pip-services3-commons-nodex";
import { MappingV1 } from "./MappingV1";

export class MappingsMemoryClientV1 implements IMappingsClient {
    private _maxPageSize: number = 100;
    private _items: MappingV1[];
    private readonly _defaultTTL: number = 7 * 24 * 60 * 60 * 1000;

    public constructor(...items: MappingV1[]) {
        this._items = items;
    }

    private composeFilter(filter: FilterParams): any {
        filter = filter || new FilterParams();
        let collection = filter.getAsNullableString("collection");
        let id = filter.getAsNullableString("id");
        let internalId = filter.getAsNullableString("internal_id");
        let externalId = filter.getAsNullableString("external_id");
        let search = filter.getAsNullableString("search");

        return (item: MappingV1) => {
            if (collection != null && item.collection != collection)
                return false;
            if (id != null && item.external_id != id && item.internal_id != id)
                return false;
            if (internalId != null && item.internal_id != internalId)
                return false;
            if (externalId != null && item.external_id != externalId)
                return false;
            if (search != null && item.external_id != search && item.internal_id != search)
                return false;
            return true;
        };
    }

    public async getCollectionNames(correlationId: string): Promise<string[]> {
        let result: string[] = [];
        for (let mapping of this._items) {
            let collection = mapping.collection;
            if (result.indexOf(collection) < 0)
                result.push(collection);
        }
        return result;
    }

    public async getMappings(correlationId: string, filter: FilterParams, paging: PagingParams): Promise<DataPage<MappingV1>> {
        let filterMappings = this.composeFilter(filter);
        let mappings = this._items.filter(filterMappings);

        // Extract a page
        paging = paging != null ? paging : new PagingParams();
        let skip = paging.getSkip(-1);
        let take = paging.getTake(this._maxPageSize);

        let total = null;
        if (paging.total)
            total = mappings.length;

        if (skip > 0)
            mappings = mappings.slice(skip);
        mappings = mappings.slice(0, take);

        let page = new DataPage<MappingV1>(mappings, total);
        return page;
    }

    public async addMapping(correlationId: string, collection: string, internalId: string, externalId: string, timeToLive: number): Promise<void> {
        let mapping: MappingV1;
        timeToLive = timeToLive > 0 ? timeToLive : this._defaultTTL;
        mapping = <MappingV1>{
            collection: collection,
            internal_id: internalId,
            external_id: externalId,
            expiration_time: new Date(new Date().getTime() + timeToLive)
        };
        if (mapping == null) {
            return;
        }
        mapping = Object.assign({}, mapping);
        mapping.id = mapping.id || IdGenerator.nextLong();
        this._items.push(mapping);
    }

    public async mapToExternal(correlationId: string, collection: string, internalId: string): Promise<string> {

        let result: string = null;

        let items = this._items.filter((m) => collection.localeCompare(m.collection) == 0 && internalId.localeCompare(m.internal_id) == 0);
        let mapping: MappingV1 = items.length > 0 ? items[0] : null;
        result = mapping != null && mapping.expiration_time > new Date() ? mapping.external_id : null;
        return result;
    }

    public async mapToInternal(correlationId: string, collection: string, externalId: string): Promise<string> {

        let result: string = null;
        let items = this._items.filter((m) => collection.localeCompare(m.collection) == 0 && externalId.localeCompare(m.external_id) == 0);
        let mapping: MappingV1 = items.length > 0 ? items[0] : null;
        result = mapping != null && mapping.expiration_time > new Date() ? mapping.internal_id : null;
        return result;
    }

    public async deleteMapping(correlationId: string, collection: string, internalId: string, externalId: string): Promise<void> {

        for (let index = this._items.length - 1; index >= 0; index--) {
            let mapping = this._items[index];
            if (mapping.collection == collection
                && mapping.internal_id == internalId
                && mapping.external_id == externalId) {
                this._items.splice(index, 1);
                break;
            }
        }
    }
}